### PR TITLE
Converted to TypeScript and enhanced CustomUI (revival)

### PR DESCRIPTION
**Pull request by sebjulliand**
### Changes
* Converted `CustomUI` to TypeScript
* Added `addXXX` methods, where `XXX` is one of the supported `Field`'s type. These methods can be chained to easily create a form.
```ts
const formPage = await new CustomUI()
  .addInput('name', 'Name')
  .addInput('address', 'Address')
  .addCheckbox('trust','Trust')
  .addHorizontalRule()
  .addButtons({id: "ok", label:"Ok"}, {id: "cancel", label:"Cancel"})
  .loadPage('My form');

if(formPage){
  formPage.panel.dispose();
  if(formPage.data){
    //Do something with data
  }
}  
```
* Updated views to use new `CustomUI` methods

### Checklist
* [x] have tested my change
* [x] updated relevant documentation
* [x] eslint is not complaining

### CODE REVIEW

**Comment by worksofliam** [comment]
@sebjulliand I am a little scared to review this one on my own. Any chance you'd be interested in meeting to go over this large one?

There's a lot to test here:

* login page
* new/edit connection page
* settings page
* actions page

**Comment by sebjulliand** [comment]
> @sebjulliand I am a little scared to review this one on my own. Any chance you'd be interested in meeting to go over this large one?
> 
> There's a lot to test here:
> 
>     * login page
> 
>     * new/edit connection page
> 
>     * settings page
> 
>     * actions page
> 
> 
> Also need to test the impact of it from the Code Coverage extension, which I believe uses this API too.

@worksofliam of course. I did a test round on all those UI pages, but you know more edge cases than I do. Let's plan something!

**Comment by sebjulliand** [comment]
@worksofliam you're good to go!

**Review by worksofliam — DISMISSED** [review]
Code reviewed, little chat with Seb about it, and tested - I'm feeling very confident about this PR.

* total cleanup
* typescript
* new addition to types API!

I also tested the Code Coverage extension with this PR and it continues to work as expected.

While I do approve this PR, @chrjorgensen, is there any chance you can have a quick test to make sure the UI works as expected for you?

**Review by chrjorgensen — COMMENTED** [review]
@sebjulliand I tried to do a thorough review - see comments below...

File: src/webviews/filters/index.js [hunk 1]
```diff
@@ -52,41 +52,14 @@ module.exports = class FiltersUI {
...
+    const ui = new CustomUI();
+    ui.addInput(`name`, `Filter name`, `The filter name should be unique.`, {default: filter.name});
+    ui.addInput(`library`, `Library`, `Library name. Cannot be generic name with an asterisk.`, {default: filter.library});
+    ui.addInput(`object`, `Object`, `Object name. Can be generic name with an asterisk. For example: <code>*</code>, or <code>Q*</code>.`, {default: filter.object});
+    ui.addInput(`types`, `Object type filter`, `A comma delimited list of object types. For example <code>*ALL</code>, or <code>*PGM, *SRVPGM</code>. <code>*SRCPF</code> is a special type which will return only source files.`, {default: filter.types.join(`, `)});
+    ui.addInput(`member`, `Member`, `Member name. Can be multi-generic value. Examples: <code>*CL</code> or <code>CL*ABC*</code>. A single <code>*</code> will return all members.`, {default: filter.member});
+    ui.addInput(`memberType`, `Member type`, `Member type. Can be multi-generic value. Examples: <code>RPG*</code> or <code>SQL*LE</code>. A single <code>*</code> will return all member types.`, {default: filter.memberType || `*`});
+    ui.addField(new Field(`submit`, `save`, `Save settings`));
```
[the remark this answers is not in the document]
Comment (worksofliam) [reply src/webviews/filters/index.js:62]: I think the old way was left, though you are right... I suppose the new version should!

File: src/views/ConnectionBrowser.ts [hunk 2]
```diff
@@ -31,10 +31,10 @@ export class ObjectBrowserProvider {
 
           switch (typeof name) {
             case `string`: // Name of connection object
-              await LoginPanel.LoginToPrevious(name, context);
+              await Login.LoginToPrevious(name, context);
```
[the remark this answers is not in the document]
Comment (sebjulliand) [reply src/views/ConnectionBrowser.ts:34]: Unless I'm wrong, the `Connect to previous` command has always had a misleading title as it doesn't let you connect to the system you were previously connected to.
It always expect an argument and when run from the command palette, it gets none. But it's been its behavior since the beginning (I went back before I converted the browser to TypeScript and it was already the case).
However, I'll create a PR to enhance it and make it work as labelled 😁

File: src/webviews/variables/index.js [hunk 3]
```diff
@@ -84,36 +69,15 @@ module.exports = class SettingsUI {
...
-        id: `delete`,
-        label: `Delete`
-      }
-    ];
-    ui.addField(field);
-
-    let {panel, data} = await ui.loadPage(`Work with Variable`);
+    const ui = new CustomUI()
+      .addInput(`name`, `Variable name`, `<code>&</code> not required. Will be forced uppercase.`, { default: currentVariable.name })
+      .addInput(`value`, `Variable value`, `<code>&</code> not required. Will be forced uppercase.`, { default: currentVariable.value })
```
[the remark this answers is not in the document]
Comment (sebjulliand) [reply src/webviews/variables/index.js:74]: Indeed!

**Review by chrjorgensen — COMMENTED** [review]
You changed it alright (even so it adds up to 100! 😆) and it looks fine to me -
but when I run the start debug command, it shows like this:
billede

The progress bar is far ahead of the progress..??? Could be an error in the VS Code API and I don't know if it's worth pursuing?

**Review by chrjorgensen — APPROVED** [review]

**Comment by sebjulliand** [comment]
> You changed it alright (even so it adds up to 100! 😆) and it looks fine to me - but when I run the start debug command, it shows like this: 
> ...
> 
> The progress bar is far ahead of the progress..??? Could be an error in the VS Code API and I don't know if it's worth pursuing?

That's odd...honestly I don't know.
Or we could just not increment on the first action, because it's the beginning of the operation, so it should not show any progress at this point.

**Comment by worksofliam** [comment]
@chrjorgensen 

> but when I run the start debug command, it shows like this:

The modal message expected behaviour, implemented by me!

**Comment by chrjorgensen** [comment]
@worksofliam It is the progress indicator, that is not correct.

@sebjulliand I don't think we should spend more time on this now. Better get the code out and maybe have a look later on...

**Comment by worksofliam** [comment]
@chrjorgensen 

image

Yep, I see your issue now. It is full here when it shouldn't be, but I believe it should now get fixed outside this PR as this PR is really supposed to be just for the webview UI changes.

Since we've goth done our reviews and are happy, let's get this merged in and then we can get another PR to fix that UI problem.

**Comment by sebjulliand** [comment]
Agreed. Go ahead and merge it. We're done here 😅

**Review by worksofliam — APPROVED** [review]
